### PR TITLE
Llama: fix custom 4D masks, v2

### PR DESCRIPTION
**Pull request by poedator**
this is an attempt to rebase  #29930 started initially by @gante 

Fixes the issue raised by @poedator in https://github.com/huggingface/transformers/pull/29753#issuecomment-2014530814.

Causal mask is now of shape [..., seq_len, full_len], as opposed to [..., full_len, full_len]. This means custom 4D attention masks are now the whole causal mask, so we don't need a sliced copy -- we can copy the whole thing :)

This PR also expands the support of custom 4D attention mask: we can pass both the full mask ([..., full_len, full_len]) or the partial mask ([..., seq_len, full_len]).
 - - - - - - - - 

as of 18.04.24 it is not passing the 4D mask tests because of the `_ignore_causal_mask_sdpa()` method, most recently edited in #30317 (merged today).

`tests/models/llama/test_modeling_llama.py::Mask4DTestHard::test_partial_stacked_causal_mask - ValueError: Incorrect 4D attention_mask shape: (1, 1, 12, 12); expected: (1, 1, 9, 12).`
apparently  `_ignore_causal_mask_sdpa()` expects that `attention_mask.shape[-2] == query_length`. This may only be true if the input_ids are contiguous, which is not always the case in some intended 4D mask applications.

`tests/models/llama/test_modeling_llama.py::Mask4DTestHard::test_stacked_causal_mask_static_cache - ValueError: Incorrect 4D attention_mask shape: (1, 1, 12, 16); expected: (1, 1, 12, 12).`
in this test  `_ignore_causal_mask_sdpa()` expects that `attention_mask.shape[-1] == key_value_length` which is set by `past seen tokens`. However, in the test I make this dimension equal to the static cache size, so that the mask always has same shape and the whole graph may be compiled and reused.

I hesitate to make edits to `_ignore_causal_mask_sdpa()` because there may be some greater context. 

Summoning @younesbelkada @ArthurZucker @gante  to help

### CODE REVIEW

**Comment by poedator** [comment]
As s a solution, I added additional `expected_shapes` to `_ignore_causal_mask_sdpa()` and improved StaticCache detection code. 
Note: it is inconvenient to have StaticCache as layer.self_attn objects and other Caches as model-level object. Perhaps there may be a model-level plug to avoid referencing the layer levels.
Please review soon - I need this for my paper code. It's been broken for quite long now.

File: src/transformers/models/cohere/modeling_cohere.py [hunk 1]
```diff
@@ -996,45 +1003,46 @@ def _update_causal_mask(
...
+                    if isinstance(attention_mask, torch.Tensor)
+                    else past_seen_tokens + sequence_length + 1
+                )
+            causal_mask = torch.full(
+                (sequence_length, target_length), fill_value=min_dtype, dtype=dtype, device=device
+            )
+            if sequence_length != 1:
+                causal_mask = torch.triu(causal_mask, diagonal=1)
+            causal_mask *= torch.arange(target_length, device=device) > cache_position.reshape(-1, 1)
+            causal_mask = causal_mask[None, None, :, :].expand(input_tensor.shape[0], 1, -1, -1)
```
Comment (ArthurZucker) [inline src/transformers/models/cohere/modeling_cohere.py:1040]: this is very messy. 
IMO if a 4D mask is passed, then we should not even touch it. 
Previously we used to invert it, now that we don't do that in the process, we should not add extra logic and just return the 4D mask. Fine with me to add a check to make sure the min is -inf, max is 0 and shape is expected, but let's not increase the surface and complicate things.

Comment (poedator) [reply src/transformers/models/cohere/modeling_cohere.py]: I simplified the code:
- removed inversion for custom 4d masks, 
- removed mask slicing for custom 4d masks
- removed test case where we pass a 4D attention mask with the full sequence length (i.e. [..., full_len, full_len]))
- updated relevant tests

I am concerned that 
- part of code related to checking masks shape sits in  `_ignore_causal_mask_sdpa()`
- the advanced 4d mask tests sit in `tests/models/llama/test_modeling_llama.py` - something may break them again outside of llama code, like in `src/transformers/modeling_attn_mask_utils.py` - could you please ensure that  `Mask4DTestHard` class is always tested if something changes in `modeling_attn_mask_utils.py` even if Llama code is not affected otherwise?

File: src/transformers/models/llama/modeling_llama.py [hunk 2]
```diff
@@ -974,13 +974,16 @@ def forward(
             inputs_embeds = self.embed_tokens(input_ids)
 
         past_seen_tokens = 0
+        is_static_cache = isinstance(past_key_values, StaticCache) or isinstance(
+            getattr(self.layers[0].self_attn, "past_key_value", None), StaticCache
+        )
```
Comment (ArthurZucker) [inline src/transformers/models/llama/modeling_llama.py:979]: Cc @gante in this release I think we can already not use any of the logic that does rely on the `DynamicCache` in favor of relying on the cache positions WDYT?

File: src/transformers/models/llama/modeling_llama.py [hunk 3]
```diff
@@ -1074,45 +1081,43 @@ def _update_causal_mask(
...
-        causal_mask *= torch.arange(target_length, device=device) > cache_position.reshape(-1, 1)
-        causal_mask = causal_mask[None, None, :, :].expand(input_tensor.shape[0], 1, -1, -1)
-        if attention_mask is not None:
-            causal_mask = causal_mask.clone()  # copy to contiguous memory for in-place edit
-            if attention_mask.dim() == 2:
+        if attention_mask is not None and attention_mask.dim() == 4:
+            # in this case we assume that the mask comes already in inverted form and requires no inversion or slicing
+            if attention_mask.max() != 0:
+                raise ValueError("Custom 4D attention mask should be passed in inverted form with max==0`")
+            causal_mask = attention_mask
```
Comment (ArthurZucker) [inline src/transformers/models/llama/modeling_llama.py:1099]: this LGTM

File: src/transformers/models/llama/modeling_llama.py [hunk 4]
```diff
@@ -1074,45 +1081,43 @@ def _update_causal_mask(
...
+                )
+            causal_mask = torch.full(
+                (sequence_length, target_length), fill_value=min_dtype, dtype=dtype, device=device
+            )
+            if sequence_length != 1:
+                causal_mask = torch.triu(causal_mask, diagonal=1)
+            causal_mask *= torch.arange(target_length, device=device) > cache_position.reshape(-1, 1)
+            causal_mask = causal_mask[None, None, :, :].expand(input_tensor.shape[0], 1, -1, -1)
+            if attention_mask is not None:
+                causal_mask = causal_mask.clone()  # copy to contiguous memory for in-place edit
```
Comment (ArthurZucker) [inline src/transformers/models/llama/modeling_llama.py:1117]: that's ugly but not the changes you brought.

File: src/transformers/models/llama/modeling_llama.py [hunk 5]
```diff
@@ -989,7 +992,10 @@ def forward(
         if position_ids is None:
             position_ids = cache_position.unsqueeze(0)
 
-        causal_mask = self._update_causal_mask(attention_mask, inputs_embeds, cache_position, past_seen_tokens)
+        static_cache_max_length = self.layers[0].self_attn.past_key_value.get_max_length() if is_static_cache else None
```
Comment (ArthurZucker) [inline src/transformers/models/llama/modeling_llama.py:995]: not super sure the static_cache_max_length should be extracted here, owuld rather hide everything in the _update_causal_mask. That's the only nit

**Review by ArthurZucker — COMMENTED** [review]
LGTM let's just not add extra checks and var that are only here for the static cache in the  forward

File: src/transformers/models/llama/modeling_llama.py (hunk 4 above)
Comment (poedator) [reply src/transformers/models/llama/modeling_llama.py:1120]: I agree with you, but fortunately it is the `else:` branch )

File: src/transformers/models/llama/modeling_llama.py (hunk 5 above)
Comment (poedator) [reply src/transformers/models/llama/modeling_llama.py]: tried to address it in [899fc6b](https://github.com/huggingface/transformers/pull/30348/commits/899fc6bd99220d2445c8cfade8372dc1aa19ae8c)

File: src/transformers/models/llama/modeling_llama.py (hunk 2 above)
Comment (gante) [reply src/transformers/models/llama/modeling_llama.py:979]: @ArthurZucker what do you mean by "logic that does rely on the DynamicCache"? The `past_seen_tokens = past_key_values.get_seq_length()` line?

**Review by gante — APPROVED** [review]
LGTM, thank you for working on it 💪 

I'm assuming all slow tests are passing for the models changed in this PR :) (if you haven't run them, please do)

**Comment by poedator** [comment]
somehow `tests/models/olmo/test_modeling_olmo.py::OlmoModelTest::test_assisted_decoding_matches_greedy_search` fails in CI, but it passes just fine locally. @gante, could you rerun this test on your side?
The OLMO code is copied from Llama if I am not mistaken.
cc @ArthurZucker

**Comment by poedator** [comment]
I noticed that mistral model support for 4D masks stayed broken after these fixes. So I added similar lines to `src/transformers/modeling_attn_mask_utils.py::_prepare_4d_causal_attention_mask_for_sdpa()`

File: src/transformers/models/llama/modeling_llama.py (hunk 2 above)
Comment (ArthurZucker) [reply src/transformers/models/llama/modeling_llama.py:979]: Yes, and the max_length as well! I just don't want to add a new variable that is not general an is only used in _update_causal_mask.

File: src/transformers/models/llama/modeling_llama.py (hunk 5 above)
Comment (ArthurZucker) [reply src/transformers/models/llama/modeling_llama.py]: A lot better thanks

**Comment by poedator** [comment]
I added `Mask4DTestHard` tests (without static cache part) to `tests/models/mistral/test_modeling_mistral.py` to ensure that the 4d masks keep working in the models that use `_prepare_4d_causal_attention_mask_for_sdpa()`. These new tests would fail without the fixes from commit d488f35 just above.
Tested the SLOW tests for `./tests/models/mistral/` branch - all fine

Is there anything left to do before the merge is possible?
@gante @ArthurZucker

File: src/transformers/models/cohere/modeling_cohere.py [hunk 6]
```diff
@@ -995,46 +998,51 @@ def _update_causal_mask(
...
-        causal_mask *= torch.arange(target_length, device=device) > cache_position.reshape(-1, 1)
-        causal_mask = causal_mask[None, None, :, :].expand(input_tensor.shape[0], 1, -1, -1)
-        if attention_mask is not None:
-            causal_mask = causal_mask.clone()  # copy to contiguous memory for in-place edit
-            if attention_mask.dim() == 2:
+        if attention_mask is not None and attention_mask.dim() == 4:
+            # in this case we assume that the mask comes already in inverted form and requires no inversion or slicing
+            if attention_mask.max() != 0:
+                raise ValueError("Custom 4D attention mask should be passed in inverted form with max==0`")
+            causal_mask = attention_mask
```
Comment (ArthurZucker) [inline src/transformers/models/cohere/modeling_cohere.py:1024]: IMO this is the only think we need!

**Review by ArthurZucker — COMMENTED** [review]
Let's remove unrelated changes!

File: src/transformers/models/cohere/modeling_cohere.py [hunk 7]
```diff
@@ -896,13 +896,16 @@ def forward(
...
+        )
         if use_cache:  # kept for BC (cache positions)
-            if not isinstance(past_key_values, StaticCache):
+            if not is_static_cache:
                 past_key_values = DynamicCache.from_legacy_cache(past_key_values)
                 past_seen_tokens = past_key_values.get_seq_length()
 
         if cache_position is None:
-            if isinstance(past_key_values, StaticCache):
+            if is_static_cache:
```
[the remark this answers is not in the document]
Comment (poedator) [reply src/transformers/models/cohere/modeling_cohere.py:908]: when StaticCache is used, the `past_key_values` argument is simply `None`, so `isinstance(past_key_values, StaticCache)` is just not working. 
As I understand it, once the StaticCache is initialized, there is no need to pass it in `past_key_values` argument. That's why getattr(self.layers[0].self_attn, "past_key_value", None) is necessary.

File: src/transformers/models/cohere/modeling_cohere.py [hunk 8]
```diff
@@ -995,46 +998,51 @@ def _update_causal_mask(
...
+                else None
+            )
+
             if AttentionMaskConverter._ignore_causal_mask_sdpa(
-                attention_mask, inputs_embeds=input_tensor, past_key_values_length=past_seen_tokens
+                attention_mask,
+                inputs_embeds=input_tensor,
+                past_key_values_length=past_seen_tokens,
+                static_cache_max_length=static_cache_max_length,
             ):
```
[the remark this answers is not in the document]
Comment (poedator) [reply src/transformers/models/cohere/modeling_cohere.py:1013]: _ignore_causal_mask_sdpa() returns `True` unless it recognize the mask shape is valid. But to do so, it needs to know the `static_cache_max_length `.

**Comment by poedator** [comment]
> Let's remove unrelated changes!

sorry, but without these changes, the fixes and tests will not work. I looked for related PRs, all I found was #30476 but it is not fixing the relevant parts of the code.

File: src/transformers/models/cohere/modeling_cohere.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (ArthurZucker) [reply src/transformers/models/cohere/modeling_cohere.py:908]: This should be fixed in another PR! 😉

File: src/transformers/models/cohere/modeling_cohere.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (ArthurZucker) [reply src/transformers/models/cohere/modeling_cohere.py:1013]: Mmmm again we should just check the attention_mask shape, if 4 just return it

**Comment by poedator** [comment]
I tried to follow Arthur's advice to streamline the path for the 4D masks and it seems to work. The relevant tests do pass. 
@ArthurZucker @gante , please review

File: src/transformers/models/llama/modeling_llama.py [hunk 9]
```diff
@@ -1080,36 +1080,44 @@ def _update_causal_mask(
...
+                            "transformers v4.42.0"
+                        )
+                        offset = cache_position[0]
+                    else:
+                        offset = 0
+                    mask_shape = attention_mask.shape
+                    mask_slice = (attention_mask.eq(0.0)).to(dtype=dtype) * min_dtype
+                    causal_mask[
+                        : mask_shape[0], : mask_shape[1], offset : mask_shape[2] + offset, : mask_shape[3]
+                    ] = mask_slice
```
Comment (ArthurZucker) [inline src/transformers/models/llama/modeling_llama.py:1120]:
```suggestion
```

File: src/transformers/models/llama/modeling_llama.py [hunk 10]
```diff
@@ -1080,36 +1080,44 @@ def _update_causal_mask(
...
-                # cache. In that case, the 4D attention mask attends to the newest tokens only.
-                if attention_mask.shape[-2] < cache_position[0] + sequence_length:
-                    logger.warning_once(
-                        "Passing a 4d mask shorter than the input length is deprecated and will be removed in "
-                        "transformers v4.42.0"
+        if attention_mask is not None and attention_mask.dim() == 4:
+            # in this case we assume that the mask comes already in inverted form and requires no inversion or slicing
+            if attention_mask.max() != 0:
+                raise ValueError("Custom 4D attention mask should be passed in inverted form with max==0`")
+            causal_mask = attention_mask
```
Comment (ArthurZucker) [inline src/transformers/models/llama/modeling_llama.py:1087]: we only treat the 4d mask here

**Review by ArthurZucker — COMMENTED** [review]
Let's remove the elif no? Looks good to me with the removed on the modeling changes!

File: src/transformers/models/llama/modeling_llama.py (hunk 9 above)
Comment (poedator) [reply src/transformers/models/llama/modeling_llama.py]: agreed and removed

File: src/transformers/models/llama/modeling_llama.py (hunk 10 above)
Comment (poedator) [reply src/transformers/models/llama/modeling_llama.py:1087]: Not sure that these lines could be deleted - this branch allows creation of causal mask and transforming it. As you suggested earlier, the custom 4D mask should come in in the right shape and in inverted form - this if block allows it to skip transformations and pass it as the causal mask.

**Comment by poedator** [comment]
I combined the 2 tests from common, which were very similar. Added tolerance - now Mixtral passes it OK.
@ArthurZucker, @gante - please see if it is good to merge now

File: src/transformers/models/bigbird_pegasus/modeling_bigbird_pegasus.py
Comment (ArthurZucker) [inline src/transformers/models/bigbird_pegasus/modeling_bigbird_pegasus.py:1]: unrelated styling change that can be removed

File: src/transformers/models/cohere/modeling_cohere.py [hunk 11]
```diff
@@ -1000,37 +1000,28 @@ def _update_causal_mask(
...
+            causal_mask = torch.full(
+                (sequence_length, target_length), fill_value=min_dtype, dtype=dtype, device=device
+            )
+            if sequence_length != 1:
+                causal_mask = torch.triu(causal_mask, diagonal=1)
+            causal_mask *= torch.arange(target_length, device=device) > cache_position.reshape(-1, 1)
+            causal_mask = causal_mask[None, None, :, :].expand(input_tensor.shape[0], 1, -1, -1)
+            if attention_mask is not None:
+                causal_mask = causal_mask.clone()  # copy to contiguous memory for in-place edit
+                if attention_mask.dim() == 2:
```
Comment (ArthurZucker) [inline src/transformers/models/cohere/modeling_cohere.py:1018]: at this point the attention mask should always be 2d no? Can probably remove the if ?

File: src/transformers/models/llama/modeling_llama.py (hunk 10 above)
Comment (ArthurZucker) [reply src/transformers/models/llama/modeling_llama.py:1087]: probably lost in the diff, this looks good to me

File: tests/models/llama/test_modeling_llama.py [hunk 12]
```diff
@@ -821,3 +822,246 @@ def test_model_7b_logits(self):
...
+        past_key_values_a = outs_1a["past_key_values"]
+
+        # Case 1: we pass a 4D attention mask regarding the current sequence length (i.e. [..., seq_len, full_len])
+        input_1b = input_ids_shared_prefix[:, part_a:]
+        position_ids_1b = position_ids_shared_prefix[:, part_a:]
+        mask_1b = mask_shared_prefix[:, :, part_a:, :]
+        outs_1b = self.model.forward(
+            input_1b, attention_mask=mask_1b, position_ids=position_ids_1b, past_key_values=past_key_values_a
+        )
+        decoded_1b = [
```
Comment (ArthurZucker) [inline tests/models/llama/test_modeling_llama.py:942]: let's try to split this to make it more readable!

File: tests/models/llama/test_modeling_llama.py [hunk 13]
```diff
@@ -821,3 +822,246 @@ def test_model_7b_logits(self):
...
+        )
+
+        outs_1b = self.model.forward(
+            input_1b,
+            attention_mask=padded_mask_1b,
+            position_ids=position_ids_1b,
+            cache_position=torch.arange(part_a, input_ids_shared_prefix.shape[-1], device=torch_device),
+            past_key_values=past_key_values,
+        )
+        decoded_1b = [
```
Comment (ArthurZucker) [inline tests/models/llama/test_modeling_llama.py:1061]: same here let's split

File: tests/models/mistral/test_modeling_mistral.py [hunk 14]
```diff
@@ -627,3 +627,127 @@ def test_speculative_generation(self):
         del model
         backend_empty_cache(torch_device)
         gc.collect()
+
+
+@slow
+@require_torch_gpu
+class Mask4DTestHard(unittest.TestCase):
```
Comment (ArthurZucker) [inline tests/models/mistral/test_modeling_mistral.py:634]: this test is the same as llama , let's use copied from

File: tests/test_modeling_common.py [hunk 15]
```diff
@@ -4195,6 +4195,80 @@ def test_flash_attn_2_from_config(self):
...
+        position_ids_shared_prefix = torch.tensor([[0, 1, 2, 3, 3, 3]], device=torch_device, dtype=torch.int64)
+
+        return input_ids, position_ids, input_ids_shared_prefix, mask_shared_prefix, position_ids_shared_prefix
+
+    def test_custom_4d_attention_mask(self):
+        if len(self.all_generative_model_classes) == 0:
+            self.skipTest("Model architecture has no generative classes, and thus not necessarily supporting 4D masks")
+
+        for model_class in self.all_generative_model_classes:
+            if not model_class._supports_cache_class:
```
Comment (ArthurZucker) [inline tests/test_modeling_common.py:4238]: brittle as one does not guarantee the other, but that's alright.

**Review by ArthurZucker — APPROVED** [review]
Thanks! 
A few small nits, but looks good to me!  
Happy that 4d support is kept 🔥

File: src/transformers/models/bigbird_pegasus/modeling_bigbird_pegasus.py
Comment (poedator) [reply src/transformers/models/bigbird_pegasus/modeling_bigbird_pegasus.py:1]: done

File: src/transformers/models/cohere/modeling_cohere.py (hunk 11 above)
Comment (poedator) [reply src/transformers/models/cohere/modeling_cohere.py:1018]: removed, tested, copied

File: tests/models/llama/test_modeling_llama.py (hunk 12 above)
Comment (poedator) [reply tests/models/llama/test_modeling_llama.py:942]: done

File: tests/models/llama/test_modeling_llama.py (hunk 13 above)
Comment (poedator) [reply tests/models/llama/test_modeling_llama.py:1061]: done

File: tests/test_modeling_common.py (hunk 15 above)
Comment (poedator) [reply tests/test_modeling_common.py:4238]: these lines came from @gante - maybe he has comments?

Comment (ArthurZucker) [reply tests/test_modeling_common.py:4238]: no worries it's alright to keep it!

**Comment by ArthurZucker** [comment]
Just waiting for the commits to be push to check the `done` parts and we can merge

File: tests/models/mistral/test_modeling_mistral.py (hunk 14 above)
Comment (poedator) [reply tests/models/mistral/test_modeling_mistral.py:634]: comparing the test code for Llama and Mistral I see that:
- specific model used in setup()
- test examples are text-based (may not lead to same token count with different tokenizers)
- Llama version has extra methods to test StaticCache

Potential solutions:
0) do nothing
1) get away from text-based examples and use token_ids for transferability across models
  1a) copy tests from Llama to Mistral
  1b) move the Mask4DTestHard class methods to testing_common

**Review by ArthurZucker — APPROVED** [review]
Nice!

File: tests/models/mistral/test_modeling_mistral.py (hunk 14 above)
Comment (ArthurZucker) [reply tests/models/mistral/test_modeling_mistral.py:634]: yes no worries
